python/pyarrow/dataset.py

```diff
@@ -310,23 +309,26 @@ def dataset(sources, filesystem=None, partitioning=None, format=None):
 
     >>> dataset("path/to/nyc-taxi/", format="parquet")
 
-    Combining different sources:
+    Construction from multiple factories:
 
     >>> dataset([
-    ...     source("s3://old-taxi-data", format="parquet"),
-    ...     source("local/path/to/new/data", format="csv")
+    ...     factory("s3://old-taxi-data", format="parquet"),
+    ...     factory("local/path/to/new/data", format="csv")
     ... ])
 
     """
-    if not isinstance(sources, list):
-        sources = [sources]
-
-    sources = [
-        _ensure_source(src, filesystem=filesystem, partitioning=partitioning,
-                       format=format)
-        for src in sources
-    ]
-    return DatasetFactory(sources).finish()
+    # bundle the keyword arguments
+    kwargs = dict(filesystem=filesystem, partitioning=partitioning,
+                  format=format)
+
+    if type(paths_or_factories) is str:
```

Comment:
Should prefer `isinstance(paths_or_factories, str)`, but I handle this in the follow-up https://github.com/bkietz/arrow/pull/1